### PR TITLE
Add "Key-Value Data" and "SQL" architecture cookbook recipes

### PR DESCRIPTION
**Pull request by miquelbeltran**
_Description of what this PR is changing or adding, and why:_

This PR follows the same pattern as https://github.com/flutter/website/pull/11394.

Adds the architecture cookbook recipes for storing key-value data and storing complex data with SQL.

The two cookbook recipes are submitted together because they share the "todo app" example under `examples/app-architecture/todo_app_service`.

cc. @ericwindmill 

_Issues fixed by this PR (if any):_

- Part of #11374

## Presubmit checklist

- [ ] This PR is marked as draft with an explanation if not meant to land until a future stable release.
- [x] This PR doesn’t contain automatically generated corrections (Grammarly or similar).
- [x] This PR follows the [Google Developer Documentation Style Guidelines](https://developers.google.com/style) — for example, it doesn’t use _i.e._ or _e.g._, and it avoids _I_ and _we_ (first person).
- [x] This PR uses [semantic line breaks](https://github.com/dart-lang/site-shared/blob/main/doc/writing-for-dart-and-flutter-websites.md#semantic-line-breaks) of 80 characters or fewer.

### CODE REVIEW

File: src/content/app-architecture/cookbook/key-value-data.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/key-value-data.md:24]:
```suggestion
If you want to learn how to store complex data on a device, 
```

File: src/content/app-architecture/cookbook/key-value-data.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/key-value-data.md:183]:
```suggestion
and handles any possible errors coming from the service layer.
```

File: src/content/app-architecture/cookbook/key-value-data.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/key-value-data.md:187]:
```suggestion
through an observable `Stream`. 
```

File: src/content/app-architecture/cookbook/key-value-data.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/key-value-data.md:191]:
```suggestion
The `ThemeRepository` depends on `SharedPreferencesService`.
```

File: src/content/app-architecture/cookbook/key-value-data.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/key-value-data.md:192]:
```suggestion
The repository obtains the stored value from the service, 
```

File: src/content/app-architecture/cookbook/key-value-data.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/key-value-data.md:193]:
```suggestion
and stores it when it changes.
```

File: src/content/app-architecture/cookbook/key-value-data.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/key-value-data.md:197]:
```suggestion receives updates.
```

File: src/content/app-architecture/cookbook/key-value-data.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/key-value-data.md:289]:
```suggestion
also create `ThemeSwitchViewModel` 
```

File: src/content/app-architecture/cookbook/sql.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/sql.md:13]:
```suggestion
might require storing data on the user’s device at some point. 
```

Google wordlist says not to use "may" unless it's a legal doc.

File: src/content/app-architecture/cookbook/sql.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/sql.md:31]:
```suggestion
which combined support for mobile and desktop. 
```

File: src/content/app-architecture/cookbook/sql.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/sql.md:33]:
```suggestion
[`sqflite_common_ffi_web`][] but it's not included in this example.
```

File: src/content/app-architecture/cookbook/sql.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/sql.md:67]:
```suggestion
- UI layer with `TodoListScreen` and `TodoListViewModel`
- Domain layer with `Todo` data class
- Data layer with `TodoRepository` and `DatabaseService`
```

File: src/content/app-architecture/cookbook/sql.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/sql.md:140]:
```suggestion
Pressing the `FilledButton`,
executes the `add` command
and passes in the text controller value.
```

File: src/content/app-architecture/cookbook/sql.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/sql.md:196]:
```suggestion
When a user taps on the `IconButton` in the `ListTile`, the delete command is executed.
```

File: src/content/app-architecture/cookbook/sql.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/sql.md:210]:
```suggestion
A correct result removes the ToDo item from the view
model *and* the screen.
```

File: src/content/app-architecture/cookbook/sql.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/sql.md:237]:
```suggestion
In this case, the application uses the `freezed` package to generate the code.
```

File: src/content/app-architecture/cookbook/sql.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/sql.md:267]:
```suggestion
You can implement the same `DatabaseService` using other storage packages 
like `sqlite3`, `drift` or even cloud storage solutions like `firebase_database`.
```

File: src/content/app-architecture/cookbook/sql.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/sql.md:339]:
```suggestion
Note that the column `id` is set as `primary key` and `autoincrement`;
```

File: src/content/app-architecture/cookbook/sql.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/sql.md:341]:
```suggestion
is assigned a new value for the `id` column.
```

File: src/content/app-architecture/cookbook/sql.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/sql.md:361]:
```suggestion
All the `DatabaseService` operations use the [`Result`][] class to return a value,
```

File: src/content/app-architecture/cookbook/sql.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/sql.md:482]:
```suggestion
[`Result`]:/app-architecture/cookbook/result
```

File: src/content/app-architecture/cookbook/sql.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/sql.md:367]:
```suggestion
For each entry, it creates a `Todo` class instance.
```

File: src/content/app-architecture/cookbook/sql.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/sql.md:414]:
```suggestion
:::note
```

File: src/content/app-architecture/cookbook/sql.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/sql.md:423]:
```suggestion
:::
```

File: src/content/app-architecture/cookbook/sql.md
Comment (sfshaza2) [inline src/content/app-architecture/cookbook/sql.md:415]:
```suggestion
In some cases, you might want to close the database when you are done with it. 
```

**Review by sfshaza2 — COMMENTED** [review]
Thanks, @miquelbeltran, for this PR! I have a question, though. These recipes haven't been added to the index file yet **and** when I tried to access them directly in the staged version, they weren't there. I don't understand that.

**Comment by miquelbeltran** [comment]
> These recipes haven't been added to the index file yet 

That's correct. I created the index in this PR: https://github.com/flutter/website/pull/11394 so I didn't add it here to avoid merge conflicts.

I'd say, we can wait until all the recipes land to create the index.

> when I tried to access them directly in the staged version, they weren't there.

They should be accessible directly:

- https://flutter-docs-prod--pr11410-mb-key-value-data-qrh7dvwm.web.app/app-architecture/cookbook/key-value-data
- https://flutter-docs-prod--pr11410-mb-key-value-data-qrh7dvwm.web.app/app-architecture/cookbook/sql

**Comment by sfshaza2** [comment]
@miquelbeltran, I'm going to go ahead and lgtm this. However, @ericwindmill or @khanhnwin need to approve the code, if they haven't already.

**Review by sfshaza2 — APPROVED** [review]
lgtm

**Review by ericwindmill — APPROVED** [review]

**Comment by miquelbeltran** [comment]
Thank you! I will rename the `Command` to `command` and then I will merge.

**Comment by miquelbeltran** [comment]
I saw the message in https://github.com/flutter/website/pull/11394#issuecomment-2489510009 so I will wait to merge until I have confirmation on where to put the cookbook recipes for app architecture.

**Comment by miquelbeltran** [comment]
Moved the files, but I will wait for #11394 to land to update the site nav

**Comment by miquelbeltran** [comment]
image

I've used the titles from the recipes, but maybe it doesn't look correct to you (maybe too long?) Any suggestions or is it fine?

**Comment by miquelbeltran** [comment]
I realized that the images on the Optimistic State cookbook were in the wrong location, I moved them in this PR if you don't mind.

**Comment by sfshaza2** [comment]
This all sounds fine, @miquelbeltran. The sidenav names are long, but are fine for now.

**Pull request merged**
